### PR TITLE
Diffusion: add JoyEcho multi-shot A/V generation support

### PR DESCRIPTION
**Pull request by niehen6174**
## Motivation

[Jdopensource/JoyAI-Echo](https://huggingface.co/jdopensource/JoyAI-Echo) (JoyEcho) is a long-form audio–video generation model built on the LTX-2 backbone. Its core idea is a **paired audio–video memory bank**: each shot commits decoded frames and audio latents into a rolling bank, and subsequent shots condition on that memory prefix — enabling **multi-shot, minute-scale generation** (officially targeting ~5-minute videos) with visual and audio continuity across prompts.

This PR adds JoyEcho support to SGLang diffusion serving:

- **Load `jdopensource/JoyAI-Echo` like any other diffusers model** — weights are materialized at runtime via [JoyAI-Echo-overlay](https://huggingface.co/Niehen6174/JoyAI-Echo-overlay), using the same overlay pattern as [LTX-2.3-overlay](https://huggingface.co/MickJ/LTX-2.3-overlay). Materialization logic lives in the overlay repo; SGLang only registers the mapping.
- **`JoyEchoPipeline` for multi-shot long video** — implements the paired audio–video memory bank so consecutive prompts in one session can share visual and audio context (see workflow below).

---

## JoyEcho vs. standard LTX-2

JoyEcho reuses much of the LTX-2 A/V stack, but inference semantics differ:

| Aspect | LTX-2.3 | JoyEcho |
|---|---|---|
| **Published weights** | Diffusers layout on HF, or monolithic + [LTX-2.3-overlay](https://huggingface.co/MickJ/LTX-2.3-overlay) | Monolithic release on HF + [JoyAI-Echo-overlay](https://huggingface.co/Niehen6174/JoyAI-Echo-overlay) |
| **Pipeline** | `LTX2Pipeline` / `LTX2TwoStageHQPipeline` | `JoyEchoPipeline` (single-stage) |
| **Denoising** | Multi-step flow matching + CFG / guider params | **LTX-2 DMD distilled path**: reuses the original LTX DMD schedule (8 steps, no flow-shift remapping), `guidance_scale=1.0` |
| **Multi-shot** | Not supported | **Paired audio–video memory bank** across shots |
| **Video VAE** | Standard LTX-2 encoder | `ltx_2_3` variant with **`ltx_2_3_condition` encoder** for memory-frame encoding |
| **Transformer** | Standard block masks | **Late-layer audio self-attention control** + memory-specific cross/self-attention masks |
| **Post-processing** | Optional two-stage HQ upscaling (pipeline-internal) | None (Multi-shot outputs per-shot mp4f) |

At a high level: JoyEcho is an LTX-2 derivative optimized for **fast distilled denoising** and **long / multi-shot generation**, not for the LTX-2.3 HQ two-stage refinement path.

---

## Multi-shot generation model

JoyEcho does **not** generate all shots in one forward pass. Each shot is one generation request; continuity is carried by an in-process **memory bank** on the pipeline instance.

**Typical workflow**

1. **Shot 0** — memory bank is empty. The model generates a standalone A/V clip.
2. **After decode** — `JoyEchoAVDecodingStage` commits the decoded video frames and packed audio latents into `PairedAudioVideoMemoryBank` (up to `memory_max_size=7` slots).
3. **Shot 1+** — `JoyEchoMemoryBankFetchStage` re-encodes stored frames through the JoyEcho VAE and fetches stored audio latents, then prepends them as a **memory prefix** before denoising.
4. **Per-shot seeding** — `JoyEchoMultishotSetupStage` applies official semantics: `prompt_seed = base_seed + shot_index`.

**Session control**

- `enable_memory_bank=True` (default): memory read/write enabled.
- `reset_memory_bank=True` (default): clears bank + shot counter at the start of a new session (`request_id` change or first shot).
- Multi-shot prompts can be passed as a **`prompt` list** (`list[str]`, one entry per shot), via `prompt_path` (`.txt`, one prompt per line), or as sequential API calls on the same server instance.

**Denoising with memory**

During `JoyEchoDMDDenoisingStage`, memory tokens are treated as **clean** (sigma = 0) prefix tokens. The stage builds:

- Paired memory **cross-attention masks** (video ↔ audio alignment per memory slot)
- Memory **self-attention block masks** (slot-internal attention only)
- **RoPE coordinate** offsets for memory vs. target segments
- **Late-layer audio mask override** (`late_layer_ratio=0.7`): upper transformer blocks disable custom audio self-attention masks to match official late-layer behavior

---

## Pipeline stages

`JoyEchoPipeline` inherits `_BaseLTX2Pipeline` and reuses standard LTX-2 front/middle stages (text encoding, connectors, AV latent prep, image encoding, timestep prep).

**JoyEcho-specific stages**

| Stage | Role |
|---|---|
| `JoyEchoMultishotSetupStage` | Multi-shot session reset; per-shot seed (`seed + shot_idx`) |
| `JoyEchoSigmaPreparationStage` | Install JoyEcho DMD sigma schedule (no LTX-2 shift remapping) |
| `JoyEchoMemoryBankFetchStage` | Encode prior-shot frames → VAE latents; load audio latents from bank; attach `joy_echo_memory` to batch |
| `JoyEchoDMDDenoisingStage` | DMD Euler denoising with optional memory prefix, custom A/V masks, late-layer audio control |
| `JoyEchoAVDecodingStage` | Standard LTX-2 A/V decode **+** commit new memory slot when `enable_memory_bank=True` |

Full stage order:

```
JoyEchoMultishotSetupStage → [LTX2 front stages] → JoyEchoSigmaPreparationStage
→ TimestepPreparationStage → LTX2AVLatentPreparationStage → LTX2ImageEncodingStage
→ JoyEchoMemoryBankFetchStage → JoyEchoDMDDenoisingStage → JoyEchoAVDecodingStage
```

---

### Core model support
- **`ltx_2_vae.py`**: `AutoencoderKLLTX2Video` supports `video_encoder_variant=ltx_2_3_condition` for JoyEcho-style memory encoding.
- **`ltx_2.py`**: Per-block `audio_self_attention_mask` override in late transformer layers (`late_layer_ratio`, `late_audio_self_attention_mask`).

### JoyEcho pipeline
- `JoyEchoPipeline`, `JoyEchoPipelineConfig`, `JoyEchoSamplingParams`
- Memory bank implementation (`PairedAudioVideoMemoryBank`) and five JoyEcho-specific stages
- Registry entry for `jdopensource/JoyAI-Echo`

### Overlay registry
- Built-in registry pin: `jdopensource/JoyAI-Echo` → `Niehen6174/JoyAI-Echo-overlay` (see Motivation; materialization details live in the overlay repo)

### CI
- `joy_echo_t2v` smoke test in `ONE_GPU_CASES` (aligned with `ltx_2_3_hq_pipeline` placement)
- Lightweight sampling: **640×384, 33 frames, 8 steps**, single-shot (`enable_memory_bank=False`)
- `run_perf_check=False`, `run_consistency_check=False` until baselines/GT are captured

---

## Accuracy Tests

**4-shot vs. official (qualitative)** — two demo videos attached (official / SGLang, 832×480 / 121 frames / 8 steps / `seed=42` / memory on). Multi-shot continuity looks reasonable, but **frames are not pixel-aligned** with official.

https://github.com/user-attachments/assets/21d4485d-2c92-4d5c-9781-5dd374a76b02

https://github.com/user-attachments/assets/4926aea6-d4b9-40f3-b3ca-1e8ec38013ba

**Why outputs differ:** With `enable_memory_bank=false`, 1-shot runs (same prompt, seed, resolution) already diverge from official in the first ~4s. That rules out memory bank and overlay weight conversion as the cause — monolithic→diffusers round-trip is lossless, and the gap appears before any memory injection.

The remaining mismatch is likely framework-level: official and SGLang sample different initial noise tensors under the same seed (layout/RNG), plus possible text-embedding, VAE-decode, and LTX-2 kernel numeric differences. Finer-grained precision alignment was not completed in this PR.

---

## Speed Tests and Profiling

vs. official JoyEcho (H100, warm model, 832×480 / 121 frames / 8 steps) — **denoise stage only**:

| Scenario | Official denoise | SGLang denoise | Speedup |
|---|---|---|---|
| 1-shot (shot 1) | 5.3s | 3.9s | **1.36×** |
| 4-shot total | 27.5s | 18.4s | **1.50×** |
| 10-shot total | 90.6s | 55.5s | **1.63×** |

Denoise advantage grows with more shots (official scales faster as memory lengthens). Memory bank caps at **7 slots** — from shot 8 onward `memory_before=7` and denoise time plateaus (~11.3s official / ~6.3s SGLang). SGLang memory fetch adds ~2.5–3.0s from shot 7.

**Decode** remains slower on SGLang (~**2.5–2.8×**; 4-shot: 5.9s vs 2.4s, 10-shot: 15.3s vs 5.5s). Timer scope vs. real path gap — TBD.

**Distributed (TP / SP):** TP is tested and works. **Ulysses SP is now supported** for both single-shot and multi-shot + memory bank.
JoyEcho uses an **asymmetric SP layout**: video target latents are time-sharded; audio (including memory tokens) is replicated on every rank so a2v/v2a cross-attention stays aligned. Implementation highlights:
- Denoising: SP-aware masks/RoPE, global target frame offsets, SP-consistent DMD renoise
- DiT: prefix-aware K/V gather — replicated memory prefix is not all-gathered; only the sharded target suffix is gathered

---

## Usage example
```bash
# Single shot
sglang generate --model-path jdopensource/JoyAI-Echo \
  --prompt "A curious raccoon" \
  --height 480 --width 832 --num-frames 121 --fps 25 \
  --num-inference-steps 8 --seed 42

# Multi-shot — prompt list in config (one string per shot; memory on by default)
cat > /tmp/joy_echo_4shot.json <<'EOF'
{
  "model_path": "jdopensource/JoyAI-Echo",
  "prompt": [
    "Shot 0 prompt ...",
    "Shot 1 prompt ...",
    "Shot 2 prompt ...",
    "Shot 3 prompt ..."
  ],
  "enable_memory_bank": true,
  "reset_memory_bank": true,
  "seed": 42,
  "height": 480,
  "width": 832,
  "num_frames": 121,
  "fps": 25
  "num_inference_steps": 8
}
EOF

sglang generate --config /tmp/joy_echo_4shot.json
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #28213113234](https://github.com/sgl-project/sglang/actions/runs/28213113234)
Latest PR Test (Extra): :x: [Run #28213113056](https://github.com/sgl-project/sglang/actions/runs/28213113056)

### CODE REVIEW

File: python/sglang/utils.py [hunk 1]
```diff
@@ -34,8 +34,7 @@
 KNOWN_NON_DIFFUSERS_DIFFUSION_MODEL_PATTERNS: dict[str, str] = {
     "hunyuan3d": "Hunyuan3D2Pipeline",
     "flux.2-dev-nvfp4": "Flux2NvfpPipeline",
-    "comfy-org/ideogram-4": "Ideogram4Nvfp4Pipeline",
-    "comfy-org--ideogram-4": "Ideogram4Nvfp4Pipeline",
+    "joy-echo": "JoyEchoPipeline",
```
Comment (mickqian) [inline python/sglang/utils.py]: do we need this? since ltx-2.3 is not here, and the weight format is similar

File: python/sglang/multimodal_gen/test/test_utils.py [hunk 2]
```diff
@@ -129,6 +129,7 @@ def roberta_processing_compat(*processor_args, **processor_kwargs):
 
 # JoyAI image editing models
 DEFAULT_JOYAI_IMAGE_EDIT_MODEL_NAME_FOR_TEST = "jdopensource/JoyAI-Image-Edit-Diffusers"
+DEFAULT_JOY_ECHO_MODEL_NAME_FOR_TEST = "jdopensource/JoyAI-Echo"
```
Comment (mickqian) [inline python/sglang/multimodal_gen/test/test_utils.py]: if it's only used once, perhaps there's no need to declare it

File: python/sglang/utils.py (hunk 1 above)
Comment (niehen6174) [reply python/sglang/utils.py]: No, this is redundant here because we have already converted to the diffusers format; it will be cleaned up.

File: python/sglang/multimodal_gen/test/test_utils.py (hunk 2 above)
Comment (niehen6174) [reply python/sglang/multimodal_gen/test/test_utils.py]: ok

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/joy_echo/denoising.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/joy_echo/denoising.py:1]: move it under `model_specific_stages/joy_echo/`

File: python/sglang/multimodal_gen/test/server/gpu_cases.py [hunk 4]
```diff
@@ -435,6 +436,19 @@
         ),
         run_component_accuracy_check=False,
     ),
+    DiffusionTestCase(
+        "joy_echo_t2v",
```
Comment (mickqian) [inline python/sglang/multimodal_gen/test/server/gpu_cases.py]: could you support sp for it, and make this testcase multi-gpu, and upload the consistency ground truth?

**Review by mickqian — COMMENTED** [review]
fantastic. please add a new cookbook section for this model

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/joy_echo/denoising.py (hunk 3 above)
Comment (niehen6174) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/joy_echo/denoising.py:1]: Ok, I'll address this later.

File: python/sglang/multimodal_gen/test/server/gpu_cases.py (hunk 4 above)
Comment (niehen6174) [reply python/sglang/multimodal_gen/test/server/gpu_cases.py]: > could you support sp for it, and make this testcase multi-gpu, and upload the consistency ground truth?

Supporting true SP in JoyEcho is a bit more involved than standard video diffusion models because of its memory mechanism and 3D attention masks. In particular:

1. Video/audio sharding must stay aligned in cross-attention.
2. RoPE requires global coordinates rather than per-rank local offsets.
3. Memory attention uses 3D `[B, Q, K]` masks, while the current SP/Ulysses path mainly assumes standard key padding masks.
4. The memory layout (replicated memory tokens + sharded target tokens) does not fit the standard SP attention workflow directly.

Would you prefer that I:
* implement full SP support in this PR, which will take some additional time, or
* first make the code compatible with SP-related parameters and multi-GPU execution, then follow up with true SP support in a separate PR?

For the testcase, I can also update it to run under multi-GPU and upload consistency ground truth accordingly.

Comment (mickqian) [reply python/sglang/multimodal_gen/test/server/gpu_cases.py]: follow-up is fine

Comment (niehen6174) [reply python/sglang/multimodal_gen/test/server/gpu_cases.py]: SP is now supported in this PR, including multi-shot + memory bank. Video is time-sharded; audio is replicated on each rank.

For CI, I added joy_echo_t2v_2gpu with consistency check enabled, and opened a GT PR: https://github.com/sgl-project/ci-data/pull/20

Could you help confirm the workflow? After ci-data#20 merges, is bumping SGL_TEST_FILES_CI_DATA_REVISION in test_utils.py and re-running CI all that’s needed, or is there anything else I should update on the sglang side?

**Comment by niehen6174** [comment]
> fantastic. please add a new cookbook section for this model

Done, added a cookbook section for the model.

File: python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py [hunk 5]
```diff
@@ -701,13 +702,38 @@ def forward(
             k = k.view(*k.shape[:-1], self.local_heads, self.dim_head)
 
             if gather_context_kv_for_sp:
-                k_full = sequence_model_parallel_all_gather(k.contiguous(), dim=1)
-                v_full = sequence_model_parallel_all_gather(v.contiguous(), dim=1)
-                gathered_mask = None
-                if mask is not None:
-                    gathered_mask = sequence_model_parallel_all_gather(
-                        mask.contiguous(), dim=1
+                # Replicated prefix (e.g. JoyEcho memory) is identical on every rank; only gather the sharded suffix.
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py:707]: could we directly use USPAttention.forward's `num_replicated_prefix` here?

Comment (niehen6174) [reply python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py:707]: I looked into using `num_replicated_prefix` here, but JoyEcho's SP layout doesn't map cleanly to the current USPAttention API:

1. **V2A cross-attn**: Q is fully replicated audio; K/V are video with `[replicated memory prefix | sharded suffix]`. This doesn't match `num_replicated_prefix` (same layout on Q/K/V) or `num_replicated_kv_prefix` (Q must be sharded). V2A also uses a 3D paired memory mask, while the replicated-prefix path in USPAttention doesn't take `attn_mask`.

2. **Masked path**: When `attn_mask is not None`, USPAttention takes the masked SP path and never reaches `num_replicated_prefix`. That path also doesn't distinguish replicated prefix from sharded suffix, so memory tokens would be incorrectly duplicated.

Fully unifying this would require extending `USPAttention` in `layer.py` (replicated-prefix + 2D/3D masks, possibly a new mode for replicated-Q cross-attn) — a shared SP infra change with broad regression risk. I'd prefer to keep the explicit gather here and track infra unification as a follow-up.

**Review by mickqian — APPROVED** [review]
